### PR TITLE
SQL-Based Player Ranks and Player Ranks Subsystem

### PR DESCRIPTION
**Pull request by GoldenAlpharex**
## About The Pull Request
So, it's been a long time coming, the player ranks (donators, mentors and veterans) have been using a file-based system to save the ckeys between rounds. That's pretty archaic and very difficult to hook anything onto.

As such, I've created a SQL-based system for storing player ranks, which should allow us to, for instance, eventually interact with player ranks from the Discord, rather than necessarily having to get in-game to have it take effect. It could also allow me to add some information to the ban reports if the player receiving the ban was a veteran or a mentor, for instance.

Not only that, I've bundled everything into a new subsystem, to make it become the one source of truth for the player ranks of players. It also allows for a lot more abstraction, making the logic for loading the ranks and adding/removing players from them a lot simpler and universal.

I've also been using some `player_rank_controller` datums, that are meant to implement all of those base interactions, just in case a player rank needs some special interactions with how it's formatted, or handled in-game.

All of this was done so that there would not need to be any further interactions with the database once the schema has been loaded, so that adding new ranks in the future would not require any more database work.

I've also added some triggers for the `player_ranks` table, which I haven't really gotten to test yet, so I really hope they'll work. If they don't, I'll make modifications until they do. The goal is that the database itself handles logging all the actions that occur on the `player_ranks` table, so that we have some cool logs to look at if need be.

~~#### Important Disclaimer: This hasn't been tested yet, because I don't have a database set up on my own test server. I'll handle testing this out on one of Skyrat's test servers instead, to validate that everything is still at the very least not instantly going to make anything crash, and that the essentials are still working.~~
It's been tested, there's only one known issue, where some people might not get their full donator status in-game. There's a workaround for it now, all the players need to do to fix it is to reconnect. Should hopefully be enough for now, until I can figure out what causes it.

There's probably some stuff that I could do better in there, I've been using protected procs and making sure that admins can't proc-call this stuff too, but I'm sure that I must've missed some spots.

### DOWNSTREAMS READ THIS:
There's a migration verb, hidden in the debug tab, and only functional for admins with `+PERMISSIONS`, `+DEBUG` and `+SERVER`, as it's ideally not a verb that should be ran more than once, even though it won't break from being ran multiple times.

If you don't want to migrate, simply make sure to add the following three entries to your `skyrat_config.txt`:
```
DONATOR_LEGACY_SYSTEM
MENTOR_LEGACY_SYSTEM
VETERAN_LEGACY_SYSTEM
```

## How This Contributes To The Skyrat Roleplay Experience
It'll allow me to do a lot more integration of the player ranks stuff both in and out of the game.

## Proof of Testing

<details>
<summary>Screenshots/Videos</summary>
  
image

It compiles, I'll have to do some more proper testing on the server (probably the test server instead of the main one) to ensure that it works properly.

</details>

## Changelog

:cl: GoldenAlpharex
refactor: Refactored the player ranks system to separate the legacy file-based system from a new SQL-based system, allowing the various systems to be interacted with far easier outside of the game.
refactor: Player ranks are now working off of a subsystem, rather than being completely disjointed, making it easier to interface with. If you're having issues with your Donator status, simply reconnect and that should fix it!
config: Added three separate configs for forcing the usage of the legacy Donator, Mentor and Veteran systems, defaulting to on (so this should hopefully not change anything on servers that have not done any migration yet).
/:cl:

### CODE REVIEW

**Comment by rosextango** [comment]
Hey Golden, I'm not sure how but there is an issue where it will suddenly remove donator status from certain players but not all. I'm not sure how to recreate it, but I know I'm not the only one experiencing it. It's happened around 12 am CST both Thursday and Friday.

**Comment by GoldenAlpharex** [comment]
> Hey Golden, I'm not sure how but there is an issue where it will suddenly remove donator status from certain players but not all. I'm not sure how to recreate it, but I know I'm not the only one experiencing it. It's happened around 12 am CST both Thursday and Friday.

For reference, this is an issue that I'm quite frankly not too sure on how to fix, unfortunately, but thankfully, I've added some code where simply reconnecting will fix that issue, for those affected.

It's an annoying issue, but I didn't want to have it delay this PR going through.

File: modular_skyrat/modules/player_ranks/code/subsystem/player_ranks.dm [hunk 1]
```diff
@@ -0,0 +1,456 @@
...
+
+
+/**
+ * Returns whether or not the user is qualified as a donator.
+ *
+ * Arguments:
+ * * user - The client to verify the donator status of.
+ */
+/datum/controller/subsystem/player_ranks/proc/is_donator(client/user)
+	if(isnull(user))
```
Comment (Useroth) [inline modular_skyrat/modules/player_ranks/code/subsystem/player_ranks.dm]: Consider adding an istype() check for client

File: modular_skyrat/modules/player_ranks/code/subsystem/player_ranks.dm [hunk 2]
```diff
@@ -0,0 +1,456 @@
...
+
+/**
+ * Returns whether or not the user is qualified as a mentor.
+ * Wrapper for the `is_mentor()` proc on the client, with a null check.
+ *
+ * Arguments:
+ * * user - The client to verify the mentor status of.
+ */
+/datum/controller/subsystem/player_ranks/proc/is_mentor(client/user)
+	if(isnull(user))
```
Comment (Useroth) [inline modular_skyrat/modules/player_ranks/code/subsystem/player_ranks.dm]: Here an istype as well

File: modular_skyrat/modules/player_ranks/code/subsystem/player_ranks.dm [hunk 3]
```diff
@@ -0,0 +1,456 @@
...
+
+
+/**
+ * Returns whether or not the user is qualified as a veteran.
+ *
+ * Arguments:
+ * * user - The client to verify the veteran status of.
+ */
+/datum/controller/subsystem/player_ranks/proc/is_veteran(client/user)
+	if(isnull(user))
```
Comment (Useroth) [inline modular_skyrat/modules/player_ranks/code/subsystem/player_ranks.dm]: And here

File: modular_skyrat/modules/player_ranks/code/subsystem/player_ranks.dm [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+	if(rank_title == donator_controller.rank_title)
+		return donator_controller
+
+	if(rank_title == mentor_controller.rank_title)
+		return mentor_controller
+
+	if(rank_title == veteran_controller.rank_title)
+		return veteran_controller
+
+	. = null // We do this to ensure that we return null.
```
Comment (Useroth) [inline modular_skyrat/modules/player_ranks/code/subsystem/player_ranks.dm]: I don't think this is necessary, given we don't assign anything else to the .

**Review by Useroth — COMMENTED** [review]
Otherwise looks pretty good to me.

**Review by Useroth — APPROVED** [review]

**Pull request merged**